Filename creation error: IllegalArgumentException: Requested character count -1 is less than zero.

**Describe the bug**
I'm seeing the following exception in CI when running `7.0` of Marathon. Not sure why CI is different than local.

**Devices (please complete the following information):**
 - Device: emulator API 30
 - OS: MacOS

**Additional context**
It also retries indefinitely for some reason.

```
[main @coroutine#59] <D.0.0.1:5037:emulator-5578]> Unknown vendor exception caught. Considering this a recoverable error
```

Stacktrace:

```
java.lang.IllegalArgumentException: Requested character count -1 is less than zero.
	at kotlin.text.StringsKt___StringsKt.take(_Strings.kt:565)
	at com.malinskiy.marathon.io.FileManager.createFilename(FileManager.kt:84)
	at com.malinskiy.marathon.io.FileManager.createFile(FileManager.kt:20)
 	at com.malinskiy.marathon.report.logs.LogWriter.saveLogs(LogWriter.kt:12)
 	at com.malinskiy.marathon.android.executor.listeners.LogCatListener.testEnded(LogCatListener.kt:42)
 	at com.malinskiy.marathon.android.executor.listeners.CompositeTestRunListener.testEnded(CompositeTestRunListener.kt:35)
 	at com.malinskiy.marathon.android.executor.listeners.CompositeTestRunListener$testEnded$1.invokeSuspend(CompositeTestRunListener.kt)
 	at (Coroutine boundary.()
 	at kotlinx.coroutines.TimeoutKt.withTimeoutOrNull(Timeout.kt:101)
 	at com.malinskiy.marathon.android.adam.AndroidDeviceTestRunner.execute(AndroidDeviceTestRunner.kt:64)
 	at com.malinskiy.marathon.android.adam.AdamAndroidDevice$execute$2$1.invokeSuspend(AdamAndroidDevice.kt:358)
 	at com.malinskiy.marathon.android.adam.AdamAndroidDevice$execute$2.invokeSuspend(AdamAndroidDevice.kt:356)
 	at com.malinskiy.marathon.android.adam.AdamAndroidDevice.execute(AdamAndroidDevice.kt:360)
 	at com.malinskiy.marathon.execution.device.DeviceActor$executeBatch$2.invokeSuspend(DeviceActor.kt:187)
 Caused by: java.lang.IllegalArgumentException: Requested character count -1 is less than zero.
 	at kotlin.text.StringsKt___StringsKt.take(_Strings.kt:565)
 	at com.malinskiy.marathon.io.FileManager.createFilename(FileManager.kt:84)
 	at com.malinskiy.marathon.io.FileManager.createFile(FileManager.kt:20)
 	at com.malinskiy.marathon.report.logs.LogWriter.saveLogs(LogWriter.kt:12)
 	at com.malinskiy.marathon.android.executor.listeners.LogCatListener.testEnded(LogCatListener.kt:42)
 	at com.malinskiy.marathon.android.executor.listeners.CompositeTestRunListener.testEnded(CompositeTestRunListener.kt:35)
 	at com.malinskiy.marathon.android.executor.listeners.CompositeTestRunListener$testEnded$1.invokeSuspend(CompositeTestRunListener.kt)
 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
 	at kotlinx.coroutines.internal.ScopeCoroutine.afterResume(Scopes.kt:33)
 	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
 	at kotlinx.coroutines.UndispatchedCoroutine.afterResume(CoroutineContext.kt:142)
 	at kotlinx.coroutines.AbstractCoroutine.resumeWith(AbstractCoroutine.kt:102)
 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
 	at java.base/java.lang.Thread.run(Thread.java:834

```

--- a/core/src/main/kotlin/com/malinskiy/marathon/io/FileManager.kt
+++ b/core/src/main/kotlin/com/malinskiy/marathon/io/FileManager.kt
@@ -81,10 +81,16 @@ class FileManager(private val output: File) {
     private fun createFilename(test: Test, fileType: FileType, limit: Int, testBatchId: String? = null): String {
         val testSuffix = "${testBatchId?.let { "-$it" } ?: ""}.${fileType.suffix}"
         val rawTestName = test.toTestName().escape()
-        val testName = rawTestName.take(limit - testSuffix.length)
+        val testName = when {
+            limit - testSuffix.length >= 0 -> rawTestName.take(limit - testSuffix.length)
+            else -> ""
+        }
         val fileName = "$testName$testSuffix"
         if (rawTestName.length > testName.length) {
-            log.error { "File name length cannot exceed $limit characters and has been trimmed to $fileName and can create a conflict. This happened because the combination of file path, test class name, and test name is too long." }
+            when {
+                limit >= 0 -> log.error { "File name length cannot exceed $limit characters and has been trimmed to $fileName and can create a conflict. This happened because the combination of file path, test class name, and test name is too long." }
+                else -> log.error { "Base path for writing a file ${rawTestName}$testSuffix is already maxed out and is ${-limit} characters more than the allowed limit of ${MAX_PATH}." }
+            }
         }
         return fileName
     }

--- a/core/src/test/kotlin/com/malinskiy/marathon/io/FileManagerTest.kt
+++ b/core/src/test/kotlin/com/malinskiy/marathon/io/FileManagerTest.kt
@@ -11,6 +11,8 @@ import com.malinskiy.marathon.test.toTestName
 import org.amshove.kluent.shouldBeEqualTo
 import org.junit.jupiter.api.Test
 import java.io.File
+import java.nio.file.Paths
+import kotlin.io.path.absolutePathString
 
 class FileManagerTest {
     private val output = Files.createTempDir()
@@ -89,5 +91,22 @@ class FileManagerTest {
             )
         )
         file.name shouldBeEqualTo "127.0.0.1-5037-emulator-5554.log"
+    }
+
+    @Test
+    fun testTooLongOutputFolder() {
+        val test = com.malinskiy.marathon.test.Test(
+            pkg = "com.xxyyzzxxyy.android.abcdefgh.abcdefghi",
+            clazz = "PackageNameTest",
+            method = "useAppContext",
+            metaProperties = emptyList()
+        )
+
+        val tempDir = Files.createTempDir()
+        val proposedPath = Paths.get(tempDir.absolutePath, FileType.LOG.name, poolId.name, deviceInfo.safeSerialNumber)
+        val additionalPathCharacters = FileManager.MAX_PATH - proposedPath.absolutePathString().length
+        val limitedOutputDirectory = File(tempDir, "x".repeat(additionalPathCharacters))
+        val limitedFileManager = FileManager(limitedOutputDirectory)
+        val file = limitedFileManager.createFile(FileType.LOG, poolId, deviceInfo, test, batchId)
     }
 }

